File: react-kanban/src/types/modals/task.ts
import { useState } from "react";
import { formatDateToString } from "../date";
import { Priority, type ColumnType, type Task } from "../serialize";
import type { TaskValidationError, UserTask } from "../validation";

export type OpenTaskModal = {
  columnType: ColumnType | null;
  task: Task | null;
};

export function openTaskModal(columnType: ColumnType | null): OpenTaskModal {
  return {
    columnType,
    task: null,
  };
}

export function openTaskModalWithTask(
  columnType: ColumnType,
  task: Task,
): OpenTaskModal {
  return {
    columnType,
    task,
  };
}

export function useTaskForm(
  task: Task | null,
  modalData: OpenTaskModal | null,
) {
  const [formModalData, setFormModalData] = useState(modalData);
  const [title, setTitle] = useState(task?.title ?? "");
  const [description, setDescription] = useState(task?.description ?? "");
  const [priority, setPriority] = useState<Priority>(
    task?.priority ?? Priority.Medium,
  );
  const [dueDate, setDueDate] = useState(
    task?.dueDate ? formatDateToString(task.dueDate) : "",
  );
  const [titleError, setTitleError] = useState<string>();
  const [priorityError, setPriorityError] = useState<string>();
  const [dueDateError, setDueDateError] = useState<string>();

  if (modalData !== formModalData) {
    setFormModalData(modalData);
    setTitle(task?.title ?? "");
    setDescription(task?.description ?? "");
    setPriority(task?.priority ?? Priority.Medium);
    setDueDate(task?.dueDate ? formatDateToString(task.dueDate) : "");
    setTitleError(undefined);
    setPriorityError(undefined);
    setDueDateError(undefined);
  }

  const userTask = (): UserTask => ({ title, description, priority, dueDate });
  const clearErrors = () => {
    setTitleError(undefined);
    setPriorityError(undefined);
    setDueDateError(undefined);
  };
  const setValidationError = (error: TaskValidationError) => {
    if (error === "TitleTooLong") {
      setTitleError("Please enter a title with at most 200 characters.");
    } else if (error === "EmptyTitle") {
      setTitleError("The title must not be empty.");
    } else if (error === "InvalidDueDate") {
      setDueDateError("Please enter a valid date.");
    } else {
      setPriorityError("Please select a valid priority.");
    }
  };

  return {
    title,
    setTitle,
    description,
    setDescription,
    priority,
    setPriority,
    dueDate,
    setDueDate,
    titleError,
    priorityError,
    dueDateError,
    userTask,
    clearErrors,
    setValidationError,
  };
}

export type TaskForm = ReturnType<typeof useTaskForm>;

File: leptos-kanban/src/types/modals/task.rs
use leptos::prelude::*;

use crate::types::{
    format_date_to_string,
    serialize::ColumnType,
    state::TaskState,
    validation::task::{TaskValidationError, UserTask},
};

#[derive(Clone)]
pub struct OpenTaskModal {
    pub column_id: Option<ColumnType>,
    pub task: Option<TaskState>,
}

impl OpenTaskModal {
    pub fn new(column_id: Option<ColumnType>) -> Self {
        Self {
            column_id,
            task: None,
        }
    }

    pub fn new_with_task(column_id: ColumnType, task: TaskState) -> Self {
        Self {
            column_id: Some(column_id),
            task: Some(task),
        }
    }
}

#[derive(Clone, Copy)]
pub struct TaskFormState {
    pub title: RwSignal<String>,
    pub description: RwSignal<String>,
    pub priority: RwSignal<String>,
    pub due_date: RwSignal<String>,
    pub title_error: RwSignal<Option<String>>,
    pub priority_error: RwSignal<Option<String>>,
    pub due_date_error: RwSignal<Option<String>>,
}

impl TaskFormState {
    pub fn new(task: Option<TaskState>) -> Self {
        let form = Self {
            title: RwSignal::new(String::new()),
            description: RwSignal::new(String::new()),
            priority: RwSignal::new(String::new()),
            due_date: RwSignal::new(String::new()),
            title_error: RwSignal::new(None),
            priority_error: RwSignal::new(None),
            due_date_error: RwSignal::new(None),
        };
        form.set_task(task);
        form
    }

    pub fn set_task(self, task: Option<TaskState>) {
        self.title.set(
            task.map(|task| task.title.get_untracked())
                .unwrap_or_default(),
        );
        self.description.set(
            task.and_then(|task| task.description.get_untracked())
                .unwrap_or_default(),
        );
        self.priority.set(
            task.map(|task| task.priority.get_untracked().to_string())
                .unwrap_or_else(|| String::from("Medium")),
        );
        self.due_date.set(
            task.and_then(|task| task.due_date.get_untracked())
                .map(|date| format_date_to_string(&date))
                .unwrap_or_default(),
        );
        self.clear_errors();
    }

    pub fn user_task(self) -> UserTask {
        UserTask {
            title: self.title.get_untracked(),
            description: self.description.get_untracked(),
            due_date: self.due_date.get_untracked(),
            priority: self.priority.get_untracked(),
        }
    }

    pub fn clear_errors(self) {
        self.title_error.set(None);
        self.priority_error.set(None);
        self.due_date_error.set(None);
    }

    pub fn set_validation_error(self, error: TaskValidationError) {
        match error {
            TaskValidationError::TitleTooLong => self.title_error.set(Some(String::from(
                "Please enter a title with at most 200 characters.",
            ))),
            TaskValidationError::EmptyTitle => self
                .title_error
                .set(Some(String::from("The title must not be empty."))),
            TaskValidationError::InvalidDueDate => self
                .due_date_error
                .set(Some(String::from("Please enter a valid date."))),
            TaskValidationError::InvalidPriority => self
                .priority_error
                .set(Some(String::from("Please select a valid priority."))),
        }
    }
}
